crates/typst/src/eval/plugin.rs

```diff
@@ -0,0 +1,204 @@
+use crate::diag::SourceResult;
+use crate::util::Bytes;
+use ecow::EcoString;
+use std::sync::Mutex;
+use typst::diag::At;
+use wasmi::{AsContext, Caller, Engine, Func as Function, Linker, Module, Value};
+
+#[derive(Debug)]
+pub struct Plugin(Mutex<PluginInstance>, Bytes);
+
+impl crate::eval::Type for Plugin {
```

Comment:
Usually, this isn't implemented manually but via
```
cast! {
    type Plugin: "plugin",
}
```
This also gives `FromValue` and `IntoValue` in addition to `Type`, which would also your `fn plugin` to directly return a `Plugin` instead of `super::Value`. The problem is that `FromValue` expects `Clone` so that might be a problem. Could be fixed with another `Arc` though.